dash/dash.py

```diff
@@ -294,28 +357,27 @@ def index(self, *args, **kwargs):  # pylint: disable=unused-argument
         scripts = self._generate_scripts_html()
         css = self._generate_css_dist_html()
         config = self._generate_config_html()
+        metas = self._generate_meta_html()
         title = getattr(self, 'title', 'Dash')
-        return '''
-        <!DOCTYPE html>
-        <html>
-            <head>
-                <meta charset="UTF-8">
-                <title>{}</title>
-                {}
-            </head>
-            <body>
-                <div id="react-entry-point">
-                    <div class="_dash-loading">
-                        Loading...
-                    </div>
-                </div>
-                <footer>
-                    {}
-                    {}
-                </footer>
-            </body>
-        </html>
-        '''.format(title, css, config, scripts)
+        if self._favicon:
+            favicon = '<link rel="icon" type="image/x-icon" href="{}">'.format(
+                flask.url_for('assets.static', filename=self._favicon))
+        else:
+            favicon = ''
+        return self.interpolate_index(
+            metas, title, css, config, scripts, _app_entry, favicon)
+
+    def interpolate_index(self,
+                          metas, title, css, config,
+                          scripts, app_entry, favicon):
+        return _interpolate(self.index_string,
```

Comment:
There's probably not enough error-checking here... Shouldn't we raise a very helpful and clear message if they forget to add `{scripts}` or something?

Reply:
Two options I thought of:
- check  in a property setter on the index_string.
  - fast only one check, but no checking on interpolate_index.
  - raise when set.
- check after interpolate_index the string contains the ids of required elements.
  - check every time the index render.
  - raise only when browsing.

Which would be best ?

Reply:
I would actually do both :) That way it fails fast for those using `index` but it still checks for those using `interpolate_index` ... better developer experience all around